ip-utils.c

```diff
@@ -0,0 +1,159 @@
+	if(ifa && query) {
+		return
+			((query->mode & IPv4OnlyMatch) && iface_matches_ipv4(ifa, &query->ipv4) == 1) ||
+			((query->mode & IPv6OnlyMatch) && iface_matches_ipv6(ifa, &query->ipv6) == 1) ||
+			((query->mode & NameOnlyMatch) && iface_matches_name(ifa, query->device_name) == 1) ? 1 : 0;
+	}
+	else {
+		return -EINVAL;
+	}
+}
+
+const struct ifaddrs * janus_query_network_devices(const struct ifaddrs * ifa, const janus_network_query_config * query) {
+	while(ifa) {
+		if(iface_matches(ifa, query) == 1) {
+			return ifa;
+		}
+		ifa = ifa->ifa_next;
+	}
+	return NULL;
+}
+
+int janus_prepare_network_device_query(const char * user_value, const JanusNetworkQueryOptions query_mode, janus_network_query_config * query) {
+	if(user_value && query) {
+		query->mode = NoMatch;
+
+		if((query_mode & IPv4OnlyMatch) && inet_pton(AF_INET, user_value, &query->ipv4) > 0) {
+			query->mode |= IPv4OnlyMatch;
+		}
+
+		if((query_mode & IPv6OnlyMatch) && inet_pton(AF_INET6, user_value, &query->ipv6) > 0) {
+			query->mode |= IPv6OnlyMatch;
+		}
+
+		if(query_mode & NameOnlyMatch) {
+			query->device_name = user_value;
+			query->mode |= NameOnlyMatch;
+		}
+
+		return ((query_mode == NoMatch) == (query->mode == NoMatch)) ? 0 : -EINVAL;
+	}
+	else {
+		return -EINVAL;
+	}
+}
+
+static int copy_ipv4(const struct sockaddr_in * iface, struct in_addr * result) {
+	result->s_addr = iface->sin_addr.s_addr;
+	return 0;
+}
+
+int janus_get_network_devices_ipv4(const struct ifaddrs * ifa, const janus_network_query_config * query, struct in_addr * result) {
+	if(ifa && (ifa->ifa_addr->sa_family == AF_INET) && result && query && (query->mode & IPv4OnlyMatch)) {
+		return copy_ipv4((struct sockaddr_in *) ifa->ifa_addr, result);
+	}
+	else {
+		return -EINVAL;
+	}
+}
+
+static int copy_ipv6(const struct sockaddr_in6 * iface, struct in6_addr * result) {
+	size_t i;
+	const uint8_t * src = iface->sin6_addr.s6_addr;
+	uint8_t * dst = result->s6_addr;
+	for(i = 0; i < 16; ++i) {
+		dst[i] = src[i];
+	}
+	return 0;
+}
+
+int janus_get_network_devices_ipv6(const struct ifaddrs * ifa, const janus_network_query_config * query, struct in6_addr * result) {
+	if(ifa && (ifa->ifa_addr->sa_family == AF_INET6) && result && query && (query->mode & IPv6OnlyMatch)) {
+		return copy_ipv6((struct sockaddr_in6 *) ifa->ifa_addr, result);
+	}
+	else {
+		return -EINVAL;
+	}
+}
+
+int janus_get_network_device_address(const struct ifaddrs * ifa, janus_network_address * result) {
+	if(ifa && result) {
+		switch(ifa->ifa_addr->sa_family) {
+			case AF_INET:
+				result->family = AF_INET;
+				return copy_ipv4((struct sockaddr_in *) ifa->ifa_addr, &result->ipv4);
+			case AF_INET6:
+				result->family = AF_INET6;
+				return copy_ipv6((struct sockaddr_in6 *) ifa->ifa_addr, &result->ipv6);
+			default:
+				return -EINVAL;
+		}
+	}
+	else {
+		return -EINVAL;
+	}
+}
+
+void janus_network_address_nullify(janus_network_address * a) {
+	if(a)
+	{
```

Comment:
Per coding style, please put the opening bracket on the same line as the `if`, in other parts of the code too if it happens.